File: crates/quilltap-core/src/doc_edit/unified_diff.rs
//! Unified diff helpers — v4 `lib/doc-edit/unified-diff.ts`.
//!
//! A real, minimal, git-style unified diff: a Myers shortest-edit script over
//! lines (see [`crate::doc_edit::line_diff`]), grouped into hunks with
//! surrounding context, so the output reads the way a `diff` is expected to read
//! (matched lines stay put, only the genuinely changed lines get `-`/`+`
//! markers, and nearby edits coalesce into a single hunk). Used for autosave
//! notifications and document-mode change summaries.
//!
//! Lines are split on `\n` only (JS `split('\n')` keeps any `\r`) and compared
//! by whole-string equality, so there is no UTF-16/offset subtlety.

use super::line_diff::{diff_lines, LineOp, LineOpType};

/// Lines of unchanged context kept on each side of a change, matching git's default.
const CONTEXT_LINES: usize = 3;

/// Guard against pathological inputs: the Myers trace is O(D · (N+M)) memory,
/// which is tiny for the common "lightly edited document" case (small D) but can
/// blow up when two very large, wholly-dissimilar files are compared. Past this
/// combined line count we fall back to a coarse whole-file replacement hunk,
/// which is still a valid diff — just not minimal.
const MAX_DIFFABLE_LINES: usize = 10000;

/// Format one side of a hunk header (`start,count`), matching git's conventions.
/// `start` is a 1-based line number (`>= 1` on every path), so `start - 1` never
/// underflows.
fn format_range(start: usize, count: usize) -> String {
    if count == 0 {
        // Empty range: git points at the line *before* the change position.
        return format!("{},0", start - 1);
    }
    if count == 1 {
        return format!("{start}");
    }
    format!("{start},{count}")
}

/// An op annotated with the 1-based line number it starts at on each side.
struct EnrichedOp<'a> {
    op: LineOp<'a>,
    old_no: usize,
    new_no: usize,
}

/// Group an edit script into unified-diff hunks, each carrying up to
/// `CONTEXT_LINES` of unchanged context on either side, coalescing changes that
/// fall within one context window of each other.
fn build_hunks(ops: &[LineOp]) -> Vec<String> {
    if !ops.iter().any(|op| op.op_type != LineOpType::Equal) {
        return Vec::new();
    }

    // Annotate each op with the 1-based line number it starts at on each side.
    let mut enriched: Vec<EnrichedOp> = Vec::with_capacity(ops.len());
    let mut old_no = 1usize;
    let mut new_no = 1usize;
    for &op in ops {
        enriched.push(EnrichedOp { op, old_no, new_no });
        match op.op_type {
            LineOpType::Equal => {
                old_no += 1;
                new_no += 1;
            }
            LineOpType::Del => old_no += 1,
            LineOpType::Ins => new_no += 1,
        }
    }

    // Locate maximal runs of changed ops, expand each by context, then merge runs
    // whose expanded ranges touch. Ranges are inclusive `(start, end)` indices.
    let mut merged: Vec<(usize, usize)> = Vec::new();
    let mut i = 0usize;
    while i < enriched.len() {
        if enriched[i].op.op_type == LineOpType::Equal {
            i += 1;
            continue;
        }
        let change_start = i;
        while i < enriched.len() && enriched[i].op.op_type != LineOpType::Equal {
            i += 1;
        }
        let change_end = i - 1;
        let start = change_start.saturating_sub(CONTEXT_LINES);
        let end = (change_end + CONTEXT_LINES).min(enriched.len() - 1);

        let extend = matches!(merged.last(), Some(last) if start <= last.1 + 1);
        if extend {
            let last = merged.last_mut().unwrap();
            last.1 = last.1.max(end);
        } else {
            merged.push((start, end));
        }
    }

    let mut hunks: Vec<String> = Vec::new();
    for (group_start, group_end) in merged {
        let slice = &enriched[group_start..=group_end];
        let old_start = slice[0].old_no;
        let new_start = slice[0].new_no;
        let mut old_count = 0usize;
        let mut new_count = 0usize;
        let mut body: Vec<String> = Vec::new();
        for e in slice {
            match e.op.op_type {
                LineOpType::Equal => {
                    old_count += 1;
                    new_count += 1;
                    body.push(format!(" {}", e.op.line));
                }
                LineOpType::Del => {
                    old_count += 1;
                    body.push(format!("-{}", e.op.line));
                }
                LineOpType::Ins => {
                    new_count += 1;
                    body.push(format!("+{}", e.op.line));
                }
            }
        }
        hunks.push(format!(
            "@@ -{} +{} @@",
            format_range(old_start, old_count),
            format_range(new_start, new_count)
        ));
        hunks.extend(body);
    }

    hunks
}

/// Coarse fallback for oversized inputs: replace the whole file in one hunk.
fn whole_file_hunk(old_lines: &[&str], new_lines: &[&str]) -> Vec<String> {
    if old_lines.is_empty() && new_lines.is_empty() {
        return Vec::new();
    }
    let mut hunks: Vec<String> = vec![format!(
        "@@ -{} +{} @@",
        format_range(1, old_lines.len()),
        format_range(1, new_lines.len())
    )];
    for line in old_lines {
        hunks.push(format!("-{line}"));
    }
    for line in new_lines {
        hunks.push(format!("+{line}"));
    }
    hunks
}

/// Generate a unified diff between two strings (v4 `generateUnifiedDiff`).
/// Produces git-style output with `@@` hunk headers and surrounding context;
/// `""` when the two inputs are identical.
pub fn generate_unified_diff(old_text: &str, new_text: &str, filename: &str) -> String {
    if old_text == new_text {
        return String::new();
    }

    // Treat truly-empty content as zero lines (git's 0-byte-file semantics) rather
    // than as `[""]`, so creating/emptying a file doesn't churn a phantom blank line.
    let old_lines: Vec<&str> = if old_text.is_empty() {
        Vec::new()
    } else {
        old_text.split('\n').collect()
    };
    let new_lines: Vec<&str> = if new_text.is_empty() {
        Vec::new()
    } else {
        new_text.split('\n').collect()
    };

    let hunks = if old_lines.len() + new_lines.len() > MAX_DIFFABLE_LINES {
        whole_file_hunk(&old_lines, &new_lines)
    } else {
        build_hunks(&diff_lines(&old_lines, &new_lines))
    };

    if hunks.is_empty() {
        return String::new();
    }

    format!("--- a/{filename}\n+++ b/{filename}\n{}", hunks.join("\n"))
}

/// Wrap a unified diff in an autosave notification (v4 `formatAutosaveNotification`).
/// `None` when the texts are identical (empty diff).
pub fn format_autosave_notification(
    old_text: &str,
    new_text: &str,
    filename: &str,
) -> Option<String> {
    let diff = generate_unified_diff(old_text, new_text, filename);
    if diff.is_empty() {
        return None;
    }
    Some(format!(
        "I've made changes to \"{filename}\":\n\n```diff\n{diff}\n```"
    ))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn identical_is_empty() {
        assert_eq!(
            generate_unified_diff("same\ntext", "same\ntext", "f.md"),
            ""
        );
        assert_eq!(generate_unified_diff("", "", "f.md"), "");
        assert_eq!(format_autosave_notification("x", "x", "f.md"), None);
    }

    #[test]
    fn minimal_hunk_context_not_churn() {
        let d = generate_unified_diff(
            "one\ntwo\nthree\nfour\nfive",
            "one\ntwo\nTHREE\nfour\nfive",
            "draft.md",
        );
        let body: Vec<&str> = d.split('\n').skip(2).collect();
        assert_eq!(
            body,
            vec![
                "@@ -1,5 +1,5 @@",
                " one",
                " two",
                "-three",
                "+THREE",
                " four",
                " five",
            ]
        );
    }

    #[test]
    fn create_from_empty_zero_old_range() {
        let d = generate_unified_diff("", "hello\nworld", "draft.md");
        assert!(d.contains("@@ -0,0 +1,2 @@"), "{d}");
        assert!(d.contains("+hello"));
        assert!(d.contains("+world"));
    }

    #[test]
    fn distant_changes_split_into_two_hunks() {
        let lines: Vec<String> = (1..=20).map(|i| format!("line{i}")).collect();
        let mut changed = lines.clone();
        changed[1] = "TOP".into();
        changed[18] = "BOTTOM".into();
        let d = generate_unified_diff(&lines.join("\n"), &changed.join("\n"), "draft.md");
        let headers = d.split('\n').filter(|l| l.starts_with("@@")).count();
        assert_eq!(headers, 2, "{d}");
    }
}

File: apps/web/src/app/documents/unified-diff.ts
/**
 * Unified-diff helpers for the Document Pane autosave notification.
 *
 * A byte-for-byte port of v4 `lib/doc-edit/line-diff.ts` (`diffLines`, a Myers
 * O(ND) shortest-edit-script) + `lib/doc-edit/unified-diff.ts`
 * (`generateUnifiedDiff` / `formatAutosaveNotification`). The client generates
 * this diff so a save can hand the server the human-readable change summary in
 * the same round-trip — the server posts it as a single Librarian save
 * announcement (v4 `handleWriteDocument`: `librarianMessage` is posted only when
 * `diffContent` is present). The in-editor change gutter's `changedBlockIndices`
 * is a tier-3 deferral and is not ported.
 *
 * @module documents/unified-diff
 */

export type LineOp = { type: 'equal' | 'del' | 'ins'; line: string };

/** Myers O(ND) shortest-edit-script diff over two arrays of lines (v4 `diffLines`). */
export function diffLines(oldLines: string[], newLines: string[]): LineOp[] {
  const n = oldLines.length;
  const m = newLines.length;

  if (n === 0 && m === 0) return [];
  if (n === 0) return newLines.map((line) => ({ type: 'ins', line }));
  if (m === 0) return oldLines.map((line) => ({ type: 'del', line }));

  const max = n + m;
  const offset = max;
  const v = new Array<number>(2 * max + 1).fill(0);
  const trace: number[][] = [];

  let found = false;
  for (let d = 0; d <= max; d++) {
    trace.push(v.slice());
    for (let k = -d; k <= d; k += 2) {
      let x: number;
      if (k === -d || (k !== d && v[k - 1 + offset] < v[k + 1 + offset])) {
        x = v[k + 1 + offset];
      } else {
        x = v[k - 1 + offset] + 1;
      }
      let y = x - k;
      while (x < n && y < m && oldLines[x] === newLines[y]) {
        x++;
        y++;
      }
      v[k + offset] = x;
      if (x >= n && y >= m) {
        found = true;
        break;
      }
    }
    if (found) break;
  }

  const ops: LineOp[] = [];
  let x = n;
  let y = m;
  for (let d = trace.length - 1; d >= 0; d--) {
    const vPrev = trace[d];
    const k = x - y;
    let prevK: number;
    if (k === -d || (k !== d && vPrev[k - 1 + offset] < vPrev[k + 1 + offset])) {
      prevK = k + 1;
    } else {
      prevK = k - 1;
    }
    const prevX = vPrev[prevK + offset];
    const prevY = prevX - prevK;

    while (x > prevX && y > prevY) {
      ops.push({ type: 'equal', line: oldLines[x - 1] });
      x--;
      y--;
    }
    if (d > 0) {
      if (x === prevX) {
        ops.push({ type: 'ins', line: newLines[y - 1] });
      } else {
        ops.push({ type: 'del', line: oldLines[x - 1] });
      }
    }
    x = prevX;
    y = prevY;
  }

  ops.reverse();
  return ops;
}

/** Lines of unchanged context kept on each side of a change (git's default). */
const CONTEXT_LINES = 3;

/** Guard against pathological inputs (Myers memory is O(D·(N+M))). */
const MAX_DIFFABLE_LINES = 10000;

/** Format one side of a hunk header (`start,count`), matching git's conventions. */
function formatRange(start: number, count: number): string {
  if (count === 0) return `${start - 1},0`;
  if (count === 1) return `${start}`;
  return `${start},${count}`;
}

type EnrichedOp = LineOp & { oldNo: number; newNo: number };

/** Group an edit script into unified-diff hunks with context (v4 `buildHunks`). */
function buildHunks(ops: LineOp[]): string[] {
  if (!ops.some((op) => op.type !== 'equal')) {
    return [];
  }

  const enriched: EnrichedOp[] = [];
  let oldNo = 1;
  let newNo = 1;
  for (const op of ops) {
    enriched.push({ ...op, oldNo, newNo });
    if (op.type === 'equal') {
      oldNo++;
      newNo++;
    } else if (op.type === 'del') {
      oldNo++;
    } else {
      newNo++;
    }
  }

  const merged: Array<{ start: number; end: number }> = [];
  let i = 0;
  while (i < enriched.length) {
    if (enriched[i].type === 'equal') {
      i++;
      continue;
    }
    const changeStart = i;
    while (i < enriched.length && enriched[i].type !== 'equal') {
      i++;
    }
    const changeEnd = i - 1;
    const start = Math.max(0, changeStart - CONTEXT_LINES);
    const end = Math.min(enriched.length - 1, changeEnd + CONTEXT_LINES);

    const last = merged[merged.length - 1];
    if (last && start <= last.end + 1) {
      last.end = Math.max(last.end, end);
    } else {
      merged.push({ start, end });
    }
  }

  const hunks: string[] = [];
  for (const group of merged) {
    const slice = enriched.slice(group.start, group.end + 1);
    const oldStart = slice[0].oldNo;
    const newStart = slice[0].newNo;
    let oldCount = 0;
    let newCount = 0;
    const body: string[] = [];
    for (const op of slice) {
      if (op.type === 'equal') {
        oldCount++;
        newCount++;
        body.push(` ${op.line}`);
      } else if (op.type === 'del') {
        oldCount++;
        body.push(`-${op.line}`);
      } else {
        newCount++;
        body.push(`+${op.line}`);
      }
    }
    hunks.push(`@@ -${formatRange(oldStart, oldCount)} +${formatRange(newStart, newCount)} @@`);
    hunks.push(...body);
  }

  return hunks;
}

/** Coarse fallback for oversized inputs: replace the whole file in one hunk. */
function wholeFileHunk(oldLines: string[], newLines: string[]): string[] {
  if (oldLines.length === 0 && newLines.length === 0) return [];
  const hunks: string[] = [
    `@@ -${formatRange(1, oldLines.length)} +${formatRange(1, newLines.length)} @@`,
  ];
  for (const line of oldLines) hunks.push(`-${line}`);
  for (const line of newLines) hunks.push(`+${line}`);
  return hunks;
}

/** Generate a git-style unified diff between two strings (v4 `generateUnifiedDiff`). */
export function generateUnifiedDiff(oldText: string, newText: string, filename: string): string {
  if (oldText === newText) {
    return '';
  }

  const oldLines = oldText === '' ? [] : oldText.split('\n');
  const newLines = newText === '' ? [] : newText.split('\n');

  const hunks =
    oldLines.length + newLines.length > MAX_DIFFABLE_LINES
      ? wholeFileHunk(oldLines, newLines)
      : buildHunks(diffLines(oldLines, newLines));

  if (hunks.length === 0) {
    return '';
  }

  return `--- a/${filename}\n+++ b/${filename}\n${hunks.join('\n')}`;
}

/**
 * Build the autosave save-announcement body (v4 `formatAutosaveNotification`):
 * a fenced diff block, or null when the two inputs are identical.
 */
export function formatAutosaveNotification(
  oldText: string,
  newText: string,
  filename: string,
): string | null {
  const diff = generateUnifiedDiff(oldText, newText, filename);
  if (!diff) {
    return null;
  }
  return `I've made changes to "${filename}":\n\n\`\`\`diff\n${diff}\n\`\`\``;
}
